contrib/server/operations/pbench-upload-results.py

```diff
@@ -0,0 +1,270 @@
+    before: Optional[datetime.datetime] = None
+
+    since_ts: Optional[float] = None
+    before_ts: Optional[float] = None
+
+    if parsed.since:
+        since = date_parser.parse(parsed.since)
+        since_ts = since.timestamp()
+
+    if parsed.before:
+        before = date_parser.parse(parsed.before)
+        before_ts = before.timestamp()
+
+    if since and before:
+        when = f" (from {since:%Y-%m-%d %H:%M} to {before:%Y-%m-%d %H:%M})"
+    elif since:
+        when = f" (since {since:%Y-%m-%d %H:%M})"
+    elif before:
+        when = f" (before {before:%Y-%m-%d %H:%M})"
+    else:
+        when = ""
+
+    if parsed.tarball.is_dir():
+        what = f"DIRECTORY {parsed.tarball}"
+    else:
+        what = f"TARBALL {parsed.tarball}"
+
+    print(f"{what}{when} -> SERVER {parsed.server}")
+
+    checkpoint: Optional[Path] = None
+    processed: list[str] = []
+    if parsed.checkpoint:
+        checkpoint = parsed.checkpoint
+        if parsed.verify:
+            print(f"Processing checkpoint state from {checkpoint}...")
+        if checkpoint.exists():
+            processed = checkpoint.read_text().splitlines()
+            if parsed.verify:
+                print(f"[CPT] done {len(processed)}")
+        if parsed.verify:
+            print(
+                f"Finished processing checkpoint data: {len(processed)} checkpointed files"
+            )
+
+    if parsed.verify:
+        print("Identifying target tarballs")
+    skipped = 0
+    early = 0
+    late = 0
+    if parsed.tarball.is_dir():
+        pool = parsed.tarball.glob("**/*.tar.xz")
+        which = []
+        for t in pool:
+            if not t.is_file():
+                continue
+            date = t.stat().st_mtime
+            if since_ts and date < since_ts:
+                early += 1
+                continue
+            if before_ts and date >= before_ts:
+                late += 1
+                continue
+            if str(t) in processed:
+                skipped += 1
+                if parsed.verify:
+                    print(f"[CPT] skip {t}")
+                continue
+            which.append(t)
+    elif parsed.tarball.is_file():
+        which = [parsed.tarball]
+    else:
+        print(f"Path {parsed.tarball} doesn't exist", file=sys.stderr)
+        return 1
+    if parsed.verify:
+        print(
+            f"Identified {len(which)} target tarballs: skipped {skipped}, {early} too old, {late} too new"
+        )
+
+    checkwriter: Optional[TextIOWrapper] = None
+    if checkpoint:
+        checkwriter = checkpoint.open(mode="a")
+
+    if parsed.verify:
+        print("Uploading target files...")
+    success = 0
+    failure = 0
+    failures = set()
+    duplicate = 0
+    for t in which:
+        try:
+            if parsed.preview:
+                print(f"UPLOAD {t}")
+                success += 1
+                if checkwriter:
+                    print(t, file=checkwriter, flush=True)
```

Comment:
If we're just "previewing", are you sure you want to count successes and, more poignantly, are you _sure_ you want to modify the checkpoint file??

Reply:
It was a useful debugging tool. For "real" use? I don't know. I'll ask the trees while I'm out hiking and see what they think ...

Reply:
The trees, oddly enough, didn't express any strong opinions. The squirrels were chattering like crazy, but don't seem to really understand the concept of "preview"; so since I found this behavior useful in testing, I'm leaving it.